buildflow/core/app/runtime/actors/runtime.py

```diff
@@ -159,55 +163,71 @@ def is_active(self):
-                        await processor_pool.actor_handle.snapshot.remote()
-                    )
-                except RayActorError:
-                    logging.exception("process actor unexpectedly died. will restart.")
-                    options = self.options.processor_options[
-                        processor_pool.processor.processor_id
-                    ]
-                    processor_pool.actor_handle = (
-                        PipelineProcessorReplicaPoolActor.remote(
-                            self.run_id,
-                            processor_pool.processor,
-                            options,
+        last_autoscale_event = datetime.utcnow()
+        # We keep running the loop while the job is running or draining to ensure
+        # we don't exit the main process before the drain is complete.
+        while (
+            self._status == RuntimeStatus.RUNNING
+            or self._status == RuntimeStatus.DRAINING
+        ):
+            # Only run the autoscale loop when the runtime is running, this prevents
+            # us from scaling while we are draining.
+            if (
+                self._status == RuntimeStatus.RUNNING
+                and last_autoscale_event + self._autoscale_frequency > datetime.utcnow()
+            ):
+                for processor_pool in self._processor_pool_refs:
+                    if self._status != RuntimeStatus.RUNNING:
+                        break
+
+                    try:
+                        processor_snapshot: ProcessorSnapshot = (
+                            await processor_pool.actor_handle.snapshot.remote()
                         )
+                    except RayActorError:
+                        logging.exception(
+                            "process actor unexpectedly died. will restart."
+                        )
+                        options = self.options.processor_options[
+                            processor_pool.processor.processor_id
+                        ]
+                        processor_pool.actor_handle = (
+                            PipelineProcessorReplicaPoolActor.remote(
+                                self.run_id,
+                                processor_pool.processor,
+                                options,
+                            )
+                        )
+                        # Restart with the default number of replicas, and let autoscale
+                        # handle the rest.
+                        processor_pool.actor_handle.run.remote()
+                        processor_pool.actor_handle.add_replicas.remote(
+                            self.options.num_replicas
+                        )
+                        continue
+                    # Updates the current backlog gauge (metric: ray_current_backlog)
+                    current_backlog = processor_snapshot.source_backlog
+                    if current_backlog is None:
+                        current_backlog = 0
```

Comment:
nit: I havent formalized this anywhere, but I think we should return -1 whenever we hit an edge case like this

Reply:
I'm going to leave it as is for now. I didn't actually touch this line git is just awful at diffs.. Just don't wanna touch too many things with the autoscaler at one time